compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -2564,7 +2569,24 @@ where
                 })
             }
             TyMaybeWithLayout::TyAndLayout(field_layout) => field_layout,
+        };
+
+        // If this field is a primitive type with the same size as the containing type
+        // (i.e., a newtype), try to preserve the valid ranges, e.g. a raw pointer field
+        // in a struct tagged with `#[rustc_layout_scalar_valid_range_start(1)]` is non-null
+        if field.fields.count() == 0 && field.size != Size::ZERO && field.size == this.layout.size()
+        {
+            if let (ty::Adt(def, _), Variants::Single { .. }) = (this.ty.kind(), &this.variants) {
+                let lcx = LayoutCx { tcx: cx.tcx(), param_env: cx.param_env() };
+                if let Abi::Scalar(scalar) = field.abi {
+                    let mut st = LayoutS::scalar(cx, scalar);
+                    lcx.adjust_scalar_valid_range(&mut st, *def);
+                    field.layout = cx.tcx().intern_layout(st);
```

Comment:
I don't think this is a bad approach, but it's likely an inefficient one - it might be interesting to keep an "effective" scalar validity range just like the "effective" alignment, in codegen.

Reply:
outdated: switched to "effective" scalar validity range approach